sled-agent/src/vmm_reservoir.rs

```diff
@@ -0,0 +1,255 @@
+
+    #[error(
+        "Communication error with VmmReservoirManager: ReplySenderDropped"
+    )]
+    ReplySenderDropped,
+}
+
+#[derive(Debug)]
+pub enum ReservoirMode {
+    None,
+    Size(u32),
+    Percentage(u8),
+}
+
+/// A message sent from [`VmmReservoirManagerHandle`] to [`VmmReservoirManager`]
+pub enum ReservoirManagerMsg {
+    SetReservoirSize {
+        mode: ReservoirMode,
+        reply_tx: oneshot::Sender<Result<(), Error>>,
+    },
+}
+
+impl ReservoirMode {
+    /// Return a configuration of the VMM reservoir as either a percentage of
+    /// DRAM or as an exact size in MiB.
+    ///
+    /// Panic upon invalid configuration
+    pub fn from_config(
+        percentage: Option<u8>,
+        size_mb: Option<u32>,
+    ) -> ReservoirMode {
+        match (percentage, size_mb) {
+            (None, None) => ReservoirMode::None,
+            (Some(p), None) => ReservoirMode::Percentage(p),
+            (None, Some(mb)) => ReservoirMode::Size(mb),
+            (Some(_), Some(_)) => panic!(
+                "only one of vmm_reservoir_percentage and \
+                vmm_reservoir_size_mb is allowed"
+            ),
+        }
+    }
+}
+
+/// A mechanism to interact with the [`VmmReservoirManager`]
+pub struct VmmReservoirManagerHandle {
+    reservoir_size: Arc<AtomicU64>,
+    tx: flume::Sender<ReservoirManagerMsg>,
+    _manager_handle: thread::JoinHandle<()>,
+}
+
+impl VmmReservoirManagerHandle {
+    /// Returns the last-set size of the reservoir
+    pub fn reservoir_size(&self) -> ByteCount {
+        self.reservoir_size.load(Ordering::SeqCst).try_into().unwrap()
+    }
+
+    /// Tell the [`VmmReservoirManager`] to set the reservoir size and wait for
+    /// a response.
+    ///
+    /// Returns an error if the allocation fails or if the manager is currently
+    /// busy handling another request.
+    //
+    // It's anticipated this will be used to change the reservation
+    #[allow(unused)]
+    pub async fn set_reservoir_size(
+        &self,
+        mode: ReservoirMode,
+    ) -> Result<(), Error> {
+        let (tx, rx) = oneshot::channel();
+        let msg = ReservoirManagerMsg::SetReservoirSize { mode, reply_tx: tx };
+        if let Err(e) = self.tx.try_send(msg) {
+            let (e, is_shutdown) = match e {
+                flume::TrySendError::Full(e) => (e, false),
+                flume::TrySendError::Disconnected(e) => (e, true),
+            };
+            let ReservoirManagerMsg::SetReservoirSize { mode, .. } = e;
+            if is_shutdown {
+                return Err(Error::Shutdown(mode));
+            } else {
+                return Err(Error::Busy(mode));
+            }
+        }
+        rx.await.map_err(|_| Error::ReplySenderDropped)?
+    }
+}
+
+/// Manage the VMM reservoir in a background thread
+pub struct VmmReservoirManager {
+    reservoir_size: Arc<AtomicU64>,
+    rx: flume::Receiver<ReservoirManagerMsg>,
+    log: Logger,
+}
+
+impl VmmReservoirManager {
+    pub fn run(
```

Comment:
nit: do this function and the next need to be `pub`? The manager definition isn't actually made visible outside of this module (if I'm reading things right the only thing you can do is get a handle to it).

I almost wonder if `VmmReservoirManager` should be private and the `spawn` routine moved to the handle impl, but I admit this is very bikesheddy.